pkg/apis/resource/types.go

```diff
@@ -0,0 +1,449 @@
+	//
+	// Setting this field is optional. If null, the resource is available
+	// everywhere.
+	AvailableOnNodes *core.NodeSelector
+
+	// ShareableResource determines whether the resource supports more
+	// than one user at a time.
+	ShareableResource bool
+}
+
+// ResourceHandleMaxSize is the maximum size of allocation.resourceHandle.
+const ResourceHandleMaxSize = 16 * 1024
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// ResourceClaimList is a collection of claims.
+type ResourceClaimList struct {
+	metav1.TypeMeta
+	// Standard list metadata
+	metav1.ListMeta
+
+	// Items is the list of resource claims.
+	Items []ResourceClaim
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// PodScheduling objects get created by a scheduler when it handles a pod which
+// uses one or more unallocated ResourceClaims with delayed allocation.
+//
+// This is an alpha type and requires enabling the DynamicResourceAllocation
+// feature gate.
+type PodScheduling struct {
+	metav1.TypeMeta
+
+	// The name must be the same as the corresponding Pod.  That Pod must
+	// be listed as owner in OwnerReferences to ensure that the
+	// PodScheduling object gets deleted when no longer needed. Normally
+	// the scheduler will delete it.
+	//
+	// Drivers must ignore PodScheduling objects where the owning pod
+	// already got deleted because such objects are orphaned and will be
+	// removed soon.
+	metav1.ObjectMeta
+
+	// Spec is set and updated by the scheduler.
+	Spec PodSchedulingSpec
+
+	// Status is updated by resource drivers.
+	Status PodSchedulingStatus
+}
+
+// PodSchedulingSpec contains the request for information about resources
+// required by a pod and eventually communicates the decision of the scheduler
+// to move ahead with pod scheduling for a specific node.
+type PodSchedulingSpec struct {
+	// When allocation is delayed, the scheduler must set the node for
+	// which it wants the resource(s) to be allocated before the driver(s)
+	// start with allocation.
+	//
+	// The driver must ensure that the allocated resource is available on
+	// this node or update PodSchedulingStatus.UnsuitableNodes to
+	// indicate where allocation might succeed.
+	//
+	// When allocation succeeds, drivers should immediately add the pod to
+	// the ResourceClaimStatus.ReservedFor field together with setting
+	// ResourceClaimStatus.Allocated. This optimization may save scheduling
+	// attempts and roundtrips through the API server because the scheduler
+	// does not need to reserve the claim for the pod itself.
+	//
+	// The selected node may change over time, for example when the initial
+	// choice turns out to be unsuitable after all. Drivers must not
+	// reallocate for a different node when they see such a change because
+	// it would lead to race conditions. Instead, the scheduler will
+	// trigger deallocation of specific claims as needed through the
+	// ResourceClaimStatus.DeallocationRequested field.
```

Comment:
This should be something like:
```golang
	// For `WaitForFirstConsumer` allocations: the name of the Node where resource
	// allocation is requested.
	//
	// For `Immediate` allocations, ??? 
	SelectedNode string
```

but the existing comment wasn't clear enough about what this should be for the immediate allocation case.

Reply:
It's not use for immediate allocation.

A lot of this text described the design, but not this particular field. I reduced the description to:
```
	// SelectedNode is the node for which allocation of ResourceClaims that
	// are referenced by the Pod and that use "WaitForFirstConsumer"
	// allocation is to be attempted.
```